ui/src/coordinator.rs

```diff
@@ -0,0 +1,1284 @@
+        // Seems not rigorous.
+        let closed = container.0.is_closed();
+        if !closed {
+            self.free_containers.push_back(container);
+        }
+
+        // Whether succeed to recycle.
+        !closed
+    }
+}
+
+#[derive(Debug, Snafu)]
+pub enum Error {
+    #[snafu(display("Reached system process limit"))]
+    SpawnWorker { source: std::io::Error },
+    #[snafu(display("Worker process's stdin not captured"))]
+    WorkerStdinCapture,
+    #[snafu(display("Worker process's stdout not captured"))]
+    WorkerStdoutCapture,
+    #[snafu(display("Failed to read child stdout"))]
+    WorkerStdoutRead { source: std::io::Error },
+    #[snafu(display("Failed to write to child stdin"))]
+    WorkerStdinWrite { source: std::io::Error },
+    #[snafu(display("Failed to deserialize worker message"))]
+    WorkerMessageDeserialization { source: bincode::Error },
+    #[snafu(display("Failed to serialize coordinator message"))]
+    CoordinatorMessageSerialization { source: bincode::Error },
+    #[snafu(display("Failed to send worker message through channel"))]
+    UnableToSendWorkerMessage {
+        source: mpsc::error::SendError<WorkerMessage>,
+    },
+    #[snafu(display("Failed to receive worker message through channel"))]
+    UnableToReceiveWorkerMessage,
+    #[snafu(display("Failed to send coordinator message through channel"))]
+    UnableToSendCoordinatorMessage {
+        source: mpsc::error::SendError<CoordinatorMessage>,
+    },
+    #[snafu(display("Failed to receive coordinator message through channel"))]
+    UnableToReceiveCoordinatorMessage,
+    #[snafu(display("Failed to convert WebSocket request to worker job"))]
+    WSRequestToWorkerJob { source: crate::Error },
+    #[snafu(display("Failed to send worker response(job report) sender through channel"))]
+    UnableToSendJobReportSender {
+        source: mpsc::error::SendError<(u128, oneshot::Sender<JobReport>)>,
+    },
+    #[snafu(display("Failed to send worker response(job report) through channel"))]
+    UnableToSendJobReport,
+    #[snafu(display("Failed to receive worker response(job report) through channel"))]
+    UnableToReceiveJobReport { source: oneshot::error::RecvError },
+    #[snafu(display("Failed to serialize playground message"))]
+    PlaygroundMessageSerialization { source: serde_json::Error },
+    #[snafu(display("Failed to deserialize WebSocket message"))]
+    WebSocketMessageDeserialization { source: serde_json::Error },
+    #[snafu(display("Failed to send playground message through channel"))]
+    UnableToSendPlaygroundMessageToWebSocket {
+        source: mpsc::error::SendError<String>,
+    },
+    #[snafu(display("WebSocket message sender closed"))]
+    WebSocketMessageSenderClosed,
+    #[snafu(display("Job report sender closed"))]
+    JobReportSenderClosed,
+}
+pub type Result<T, E = Error> = ::std::result::Result<T, E>;
+
+// It's safe unless a leap second happens.
+fn now_as_uid() -> u128 {
```

Comment:
My thought was that using a shared counter increases contention.
Maybe time is not bad.

Reply:
The other question is why do we need a unique ID in the first place? Often times, that can be driven out of the system at some higher level.

I don't fully follow it here, but what would happen if it was just a local variable? I've stripped out a lot of intermediate code to show the general idea:

```rust
    let mut outer_id = 0; // local
    loop {        
        tokio::select! {
            ws_request = ws_receiver.recv() => {
                if let Some(txt) = ws_request {
                    let ws_msg = serde_json::from_str(&txt).context(WebSocketMessageDeserializationSnafu);
                    match ws_msg {
                        WSMessage::Request(ws_request) => {
                            let id = outer_id; // we need one, so use and increment
                            outer_id += 1;
                            current_job = Some(tokio::spawn(async move {
                                let coordinator_msg = CoordinatorMessage::Request(id, job);
```

Reply:
We need a unique ID for each high-level request so that we can aggregate low-level operation results into response of correct kind - `ExecuteResponse` vs `CompileResponse`.

Reply:
You're right we could just use a local integer.

Reply:
> so that we can aggregate low-level operation results

> we could just use a local integer

Perfect, that's what I was hoping.